File: square-root/square-root.py
def square(x): return x * x
def average(a, b): return (a + b) / 2


def square_root(x):
  def good_enough(guess):
    return abs(square(guess) - x) < 0.001

  def improve(guess):
    return average(guess, x / guess)

  guess = 1
  while not good_enough(guess):
    guess = improve(guess)
  return guess


print(square_root(25))
print(square_root(49))
print(square_root(square(10)))

File: square-root/square-root.js
const print = console.log;
const square = (x) => x * x;
const abs = (x) => x >= 0 ? x : -x;
const average = (a, b) => (a + b) / 2;

function square_root(x) {
  const good_enough = (guess) => abs(square(guess) - x) < 0.001;
  const improve = (guess) => average(guess, x / guess);

  let guess = 1;
  while (!good_enough(guess)) {
    guess = improve(guess);
  }
  return guess;
}

print(square_root(25));
print(square_root(49));
print(square_root(square(10)));
